sklearn/utils/testing.py

```diff
@@ -390,20 +394,83 @@ def __exit__(self, *exc_info):
     assert_greater = _assert_greater
 
 
+if hasattr(np.testing, 'assert_allclose'):
+    assert_allclose = np.testing.assert_allclose
+else:
+    assert_allclose = _assert_allclose
+
+
+def assert_safe_sparse_allclose(val1, val2, rtol=1e-7, atol=0, msg=None):
+    """Check if two objects are close up to the preset tolerance.
+
+    The objects can be scalars, lists, tuples, ndarrays or sparse matrices.
+    """
+    if msg is None:
+        msg = ("The val1,\n%s\nand val2,\n%s\nare not all close"
+               % (val1, val2))
+
+    if isinstance(val1, str) and isinstance(val2, str):
+        assert_true(val1 == val2, msg=msg)
+
+    elif np.isscalar(val1) and np.isscalar(val2):
+        assert_allclose(val1, val2, rtol=rtol, atol=atol, err_msg=msg)
+
+    # To allow mixed formats for sparse matrices alone
+    elif type(val1) is not type(val2) and not (
+            sp.issparse(val1) and sp.issparse(val2)):
+        assert False, msg
+
+    elif not (isinstance(val1, (list, tuple, np.ndarray, sp.spmatrix, dict))):
+        raise ValueError("The objects,\n%s\nand\n%s\n, are neither scalar nor "
+                         "array-like." % (val1, val2))
+
+    # list/tuple/dict (of list/tuple/dict...) of ndarrays/spmatrices/scalars
+    elif isinstance(val1, (tuple, list, dict)):
+        if isinstance(val1, dict):
+            val1, val2 = tuple(val1.iteritems()), tuple(val2.iteritems())
+        if (len(val1) == 0) and (len(val2) == 0):
+            assert True
+        elif len(val1) != len(val2):
+            assert False, msg
+        # nested lists/tuples - [array([5, 6]), array([5, ])] and [[1, 3], ]
+        # Or ['str',] and ['str',]
+        elif isinstance(val1[0], (tuple, list, np.ndarray, sp.spmatrix, str)):
+            # Compare them recursively
+            for i, val1_i in enumerate(val1):
+                assert_safe_sparse_allclose(val1_i, val2[i],
+                                            rtol=rtol, atol=atol, msg=msg)
+        # Compare the lists using np.allclose, if they are neither nested nor
+        # contain strings
+        else:
+            assert_allclose(val1, val2, rtol=rtol, atol=atol, err_msg=msg)
+
+    # scipy sparse matrix
+    elif sp.issparse(val1) or sp.issparse(val2):
+        # NOTE: ref np.allclose's note for assymetricity in this testing
+        if val1.shape != val2.shape:
+            assert False, msg
+
+        diff = abs(val1 - val2) - (rtol * abs(val2))
+        assert np.any(diff > atol).size == 0, msg
+
+    # numpy ndarray
+    elif isinstance(val1, (np.ndarray)):
+        if val1.shape != val2.shape:
+            assert False, msg
+        assert_allclose(val1, val2, rtol=rtol, atol=atol, err_msg=msg)
+    else:
+        assert False, msg
```

Comment:
@vene @jnothman Could you look at this implementation once? (This is still WIP as 30% of the tests (fit reset tests) don't pass... but I'd like to know if I am going in the right direction)